### PR TITLE
eth, miner: add timeout for building sealing block

### PR DESCRIPTION
**Pull request by rjl493456442**
This PR adds a timeout mechanism for building sealing block. 

Once The Merge is conducted, block has to be prepared in the period of a slot. If we encounter some transactions that perform particularly expensive, instead of spending a lot of time filling the block, we would rather return a block that is not full, but it can be created at the specified time.

The timeout value should be chosen aligned with consensus spec.

TODO 

- [x] Make the timeout configurable via CLI flag

### CODE REVIEW

File: eth/catalyst/api.go [hunk 1]
```diff
@@ -22,6 +22,7 @@ import (
 	"encoding/binary"
 	"errors"
 	"fmt"
+	"github.com/ethereum/go-ethereum/eth/ethconfig"
```
Comment (holiman) [inline eth/catalyst/api.go:25]: goimports

File: eth/catalyst/api.go [hunk 2]
```diff
@@ -95,14 +103,20 @@ type ConsensusAPI struct {
...
-func NewConsensusAPI(eth *eth.Ethereum) *ConsensusAPI {
+func NewConsensusAPI(eth *eth.Ethereum, newpayloadTimeout time.Duration) *ConsensusAPI {
 	if eth.BlockChain().Config().TerminalTotalDifficulty == nil {
 		log.Warn("Engine API started but chain not configured for merge yet")
 	}
+	// Sanitize the timeout config for creating payload.
+	if newpayloadTimeout < time.Millisecond*100 {
+		log.Warn("Sanitizing new payload timeout to default", "provided", newpayloadTimeout, "updated", ethconfig.Defaults.Miner.NewPayloadTimeout)
+		newpayloadTimeout = ethconfig.Defaults.Miner.NewPayloadTimeout
+	}
```
Comment (holiman) [inline eth/catalyst/api.go:114]: Isn't it a bit odd if `101ms` -> `101ms`, but `99ms` -> `2s` ? 
How about something like this?
```suggestion
	if newpayloadTimeout == 0 {
		log.Warn("Sanitizing new payload timeout to default", "provided", newpayloadTimeout, "updated", ethconfig.Defaults.Miner.NewPayloadTimeout)
		newpayloadTimeout = ethconfig.Defaults.Miner.NewPayloadTimeout
	}
	if newpayloadTimeout < time.Millisecond*100 {
		log.Warn("Low payload timeout may cause high amount of empty blocks", "provided", newpayloadTimeout, "default", ethconfig.Defaults.Miner.NewPayloadTimeout)
	}
```
If someone does decide to set very small timeout causing empty blocks, well, it's their node...

**Review by holiman — COMMENTED** [review]
Looks mostly good to me

File: eth/catalyst/api.go (hunk 2 above)
Comment (rjl493456442) [reply eth/catalyst/api.go:114]: yup, good idea

**Review by holiman — COMMENTED** [review]
All in all, I think this change is fine. The one thing I feel I need to investigate a bit more, is how you re-wired the resubmit adjustments. You moved it out from commitTransactions, to `fillTransactions`. Which I think is fine. But I believe it could also be triggered by the `mainLoop` before, which it cannot now?

**Comment by holiman** [comment]
> But I believe it could also be triggered by the mainLoop before, which it cannot now?

Ah, so that case is fine: it could be triggered from the mainLoop, but only if we're not actively mining. 
```
			// Apply transactions to the pending state if we're not sealing
			//
			// Note all transactions received may not be continuous with transactions
			// already included in the current sealing block. These transactions will
			// be automatically eliminated.
			if !w.isRunning() && w.current != nil {

```
If we're not actively mining, it doesn't matter about tuning the resubmit interval.

**Review by holiman — APPROVED** [review]
LGTM

**Comment by rjl493456442** [comment]
@holiman the recommit interval adjustment can still be triggered in `mainLoop` whenever we serve the requests for creating new mining block. Specifically in line 553, worker.go. In line 619, this logic will never be triggered since the interruption signal passed is nil. All in all, it's just a code movement shouldn't touch the original logic.

But honestly since we have switched to PoS, I guess we should retire block building logic in miner and can only build block via catalyst.

**Pull request merged**